File: All sorting/merge sort/merge java/MergeSort1000.java
import java.util.*;

public class MergeSort1000 {

    static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;

            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            merge(arr, left, mid, right);
        }
    }

    static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++)
            L[i] = arr[left + i];

        for (int j = 0; j < n2; j++)
            R[j] = arr[mid + 1 + j];

        int i = 0, j = 0, k = left;

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    static void printArray(int[] arr) {
        System.out.println(Arrays.toString(arr));
    }

    public static void main(String[] args) {

        int n = 1000;
        Random rand = new Random();

        int[] randomCase = new int[n];
        for (int i = 0; i < n; i++)
            randomCase[i] = rand.nextInt(200) + 1;

        int[] ascendingCase = new int[n];
        for (int i = 0; i < n; i++)
            ascendingCase[i] = i + 1;

        int[] descendingCase = new int[n];
        for (int i = 0; i < n; i++)
            descendingCase[i] = n - i;

        int[] partialCase = ascendingCase.clone();
        for (int i = 50; i < n; i++) {
            int j = 50 + rand.nextInt(n - 50);
            int temp = partialCase[i];
            partialCase[i] = partialCase[j];
            partialCase[j] = temp;
        }

        int[] missingCase = new int[n];
        for (int i = 0; i < n; i++)
            missingCase[i] = rand.nextInt(150) + 1;

        int[] duplicateCase = new int[n];
        for (int i = 0; i < n; i++)
            duplicateCase[i] = rand.nextInt(20) + 1;

        String[] names = {
                "Random",
                "Ascending",
                "Descending",
                "Partial Order",
                "Missing Numbers",
                "Duplicates"
        };

        int[][] cases = {
                randomCase,
                ascendingCase,
                descendingCase,
                partialCase,
                missingCase,
                duplicateCase
        };

        for (int i = 0; i < cases.length; i++) {

            int[] arr = cases[i].clone();

            System.out.println("\n======================================");
            System.out.println("Case: " + names[i]);

            System.out.println("Original Array:");
            printArray(arr);

            long start = System.nanoTime();

            mergeSort(arr, 0, arr.length - 1);

            long end = System.nanoTime();

            System.out.println("Sorted Array:");
            printArray(arr);

            System.out.printf("Execution Time: %.6f ms%n",
                    (end - start) / 1000000.0);
        }
    }
}

File: All sorting/merge sort/merge c++/mergesort_1000.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <chrono>

using namespace std;
using namespace chrono;

void merge(vector<int> &arr, int left, int mid, int right) {
    vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);

    int i = 0, j = 0, k = left;

    while (i < L.size() && j < R.size()) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < L.size())
        arr[k++] = L[i++];

    while (j < R.size())
        arr[k++] = R[j++];
}

void mergeSort(vector<int> &arr, int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;

        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        merge(arr, left, mid, right);
    }
}

void printArray(vector<int> arr) {
    for (int x : arr)
        cout << x << " ";
    cout << endl;
}

int main() {

    srand(time(0));

    int n = 1000;

    vector<int> randomCase, ascendingCase, descendingCase, partialCase, missingCase, duplicateCase;

    // Random Numbers
    for (int i = 0; i < n; i++)
        randomCase.push_back(rand() % 200 + 1);

    // Ascending Order
    for (int i = 1; i <= n; i++)
        ascendingCase.push_back(i);

    // Descending Order
    for (int i = n; i >= 1; i--)
        descendingCase.push_back(i);

    // Partial Order
    partialCase = ascendingCase;
    random_shuffle(partialCase.begin() + 50, partialCase.end());

    // Missing Numbers
    for (int i = 0; i < n; i++)
        missingCase.push_back(rand() % 150 + 1);

    // Duplicates
    for (int i = 0; i < n; i++)
        duplicateCase.push_back(rand() % 20 + 1);

    string names[] = {
        "Random",
        "Ascending",
        "Descending",
        "Partial Order",
        "Missing Numbers",
        "Duplicates"
    };

    vector<vector<int>> cases = {
        randomCase,
        ascendingCase,
        descendingCase,
        partialCase,
        missingCase,
        duplicateCase
    };

    for (int i = 0; i < cases.size(); i++) {

        vector<int> arr = cases[i];

        cout << "\n==========================================" << endl;
        cout << "Case: " << names[i] << endl;

        cout << "Original Array:" << endl;
        printArray(arr);

        auto start = high_resolution_clock::now();

        mergeSort(arr, 0, arr.size() - 1);

        auto end = high_resolution_clock::now();

        cout << "Sorted Array:" << endl;
        printArray(arr);

        double timeTaken = duration<double, milli>(end - start).count();

        cout << "Execution Time: " << timeTaken << " ms" << endl;
    }

    return 0;
}